psydac/core/bsplines_pyccel.py

```diff
@@ -0,0 +1,1018 @@
+
+    Parameters
+    ----------
+    breaks : array_like
+        Coordinates of breakpoints (= cell edges); given in increasing order and
+        with no duplicates.
+
+    degree : int
+        Spline degree (= polynomial degree within each interval).
+
+    periodic : bool
+        True if domain is periodic, False otherwise.
+
+    multiplicity: int
+        Multiplicity of the knots in the knot sequence, we assume that the same
+        multiplicity applies to each interior knot.
+
+    out : array
+        The result will be inserted into this array.
+        It should be of the appropriate shape and dtype.
+.
+    """
+    for i in range(1, len(breaks) - 1):
+        out[degree + 1  + (i - 1) * multiplicity:degree + 1 + i * multiplicity] = breaks[i]
+
+    out[degree] = breaks[0]
+    out[len(out) - degree - 1] = breaks[-1]
+
+    if periodic:
+        period = breaks[-1]-breaks[0]
+        for i in range(degree):
+            out[i] = breaks[len(breaks) - degree - 1 + i] - period
+            out[len(out) - 1 - i] = breaks[degree - i] + period
+    else:
+        out[0:degree + 1] = breaks[0]
+        out[len(out) - degree - 1:] = breaks[-1]
+
+
+# =============================================================================
+def elevate_knots_p(knots: 'float[:]', degree: int, periodic: bool, out: 'float[:]',
+                    multiplicity: int = 1,
+                    tol: float = 1e-15):
+    """
+    Given the knot sequence of a spline space S of degree p, compute the knot
+    sequence of a spline space S_0 of degree p+1 such that u' is in S for all
+    u in S_0.
+
+    Specifically, on bounded domains the first and last knots are repeated in
+    the sequence, and in the periodic case the knot sequence is extended by
+    periodicity.
+
+    Parameters
+    ----------
+    knots : array_like
+        Knots sequence of spline space of degree p.
+
+    degree : int
+        Spline degree (= polynomial degree within each interval).
+
+    periodic : bool
+        True if domain is periodic, False otherwise.
+
+    multiplicity : int
+        Multiplicity of the knots in the knot sequence, we assume that the same
+        multiplicity applies to each interior knot.
+
+    tol: float
+        If the distance between two knots is less than tol, we assume
+        that they are repeated knots which correspond to the same break point.
+
+    out : array
+        The result will be inserted into this array.
+        It should be of the appropriate shape and dtype.
+    """
+
+    if periodic:
+        T, p = knots, degree
+        period = T[len(knots) -1 - p] - T[p]
+        left   = T[len(knots) -2 - 2 * p] - period
+        right  = T[2 * p + 1] + period
+
+        out[0] = left
+        out[-1] = right
+        out[1: - 1] = knots
+
+    else:
+        out[0] = knots[0]
+        out[1:degree + 2] = knots[:degree+1]
+
+        n_out = out.shape[0]
+        n_knots = len(knots)
```

Comment:
Why do you use `shape[0]` in the first case, and `len()` in the second?

Reply:
Used `len` in both case in 0aa50ba